README.txt
========================================================
 feedcache - Maintain a cache of RSS and Atom Feed Data
========================================================

The feedcache package implements a class to wrap Mark Pilgrim's
Universal Feed Parser module so that parameters can be used to cache
the feed results locally instead of fetching the feed every time it is
requested. Uses both etag and modified times for caching. The cache is
parameterized to use different backend storage options.

Compatible with Python 3

ChangeLog
2022-08-20 Alex Gheorghita <alex@agilegeeks.ro>
	* ported code to work with python3
	* added Dockerfile to support library development under Docker

2007-11-25  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/test_server.py (TestHTTPServer.__init__): Initialize
	the expected response.

	* feedcache/cache.py (Cache.purge): Add a method to clear old
	content from the cache.

	* feedcache/test_server.py (TestHTTPHandler): Use a separate
	method for handling each response code so we can have different
	behaviors based on what the test wants returned.

	* feedcache/test_cache.py (CacheRedirectHandlingTest): Add tests
	for redirect handling.

	* feedcache/cache.py (Cache.fetch): Update docstring to explain
	how redirected URLs are handled.

2007-09-02  Doug Hellmann  <doug.hellmann@gmail.com>

	* README.txt: Add README.txt and LICENSE.txt.

2007-08-21  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/cache.py (Cache.fetch): Add force_update and offline
	flags based on patch from Thomas Perl.

2007-08-08  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/example.py (main): Use Cache.fetch() instead of the
	old Cache.__getitem__().

2007-08-07  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/test_server.py (TestHTTPHandler.do_GET): Clean up the
	etag and modified date logic.

2007-08-06  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/cache.py (Cache.fetch): Change __getitem__ to fetch
	since Cache does not support the rest of the dictionary API.

	* feedcache/example.py: Remove logging configuration so log
	messages are discarded quietly.

	* feedcache/test_cache.py: Use dictionary instead of custom
	MemoryStorage class.

	* feedcache/cache.py (Cache.__getitem__): Use __getitem__ and
	__setitem__ API with storage.

2007-08-05  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/storagebase.py (StorageBase.get): Merge getContent and
	getModifiedTime into a single method, get, to avoid race
	conditions between checking the time and using the data.

	* feedcache/test_shelvestorage.py (ShelveStorageCacheTest): Add a
	few integration tests using ShelveStorage with the Cache.

	* feedcache/shelvestorage.py: Check that the storage is open
	before allowing the caller to use it.  Handle KeyError exceptions
	by returning None.

	* feedcache/example.py (main): Simple example program that uses
	the feed cache to retrieve feed data and print out the titles.

2007-08-04  Doug Hellmann  <doug.hellmann@gmail.com>

	* feedcache/test_server.py (TestHTTPHandler.do_GET): Only check
	the ETag and If-Modified-Since header values if the server is
	configured to apply them.

	* feedcache/test_cache.py: Make all of the tests use the HTTP
	server, and add a few additional checks.

	* feedcache/cache.py (Cache.__getitem__): Only check the age of
	the item in the cache if we have a time-to-live value.

	* setup.py.in: Add requires=['feedparser'].

	* feedcache/cache.py (Cache): Add URL to feedparser.org in docstring.

	* feedcache/__init__.py: Import useful classes directly to the
	package level.

	* feedcache/test_shelvestorage.py (ShelveStorageTest.testRetrieveExistingData):
	Ensure that the shelvestorage is saving modified times.

	* feedcache/test_server.py (TestHTTPHandler.do_GET): Include the
	If-Modified-Since and Last-Modified header handling.

	* feedcache/cache.py (Cache.__getitem__): Only store the newly
	fetched feed data in the cache storage if the status code is not
	304 and there was no error.

	* test_server.py (TestHTTPServer): HTTP server for tests.
	(TestHTTPHandler): HTTP request handler for tests.

	* test_cache.py: Set up the logging before anything else.
	(CacheServerTest): New tests that work with an HTTP server to
	check etag handling in the cache.

	* memorystorage.py (MemoryStorage.getModifiedTime): Change getAge
	to getModifiedTime.

	* cache.py (Cache.__getitem__): Change getAge to getModifiedTime.
	(Cache): Add debugging.



feedcache/test_cache.py
#!/usr/bin/env python
#
# Copyright 2007 Doug Hellmann.
#
#
#                         All Rights Reserved
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby
# granted, provided that the above copyright notice appear in all
# copies and that both that copyright notice and this permission
# notice appear in supporting documentation, and that the name of Doug
# Hellmann not be used in advertising or publicity pertaining to
# distribution of the software without specific, written prior
# permission.
#
# DOUG HELLMANN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
# NO EVENT SHALL DOUG HELLMANN BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
# OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""Unittests for feedcache.cache

"""

__module_id__ = "$Id$"

import logging
logging.basicConfig(level=logging.DEBUG,
                    format='%(asctime)s %(levelname)-8s %(name)s %(message)s',
                    )
logger = logging.getLogger('feedcache.test_cache')

#
# Import system modules
#
import copy
import time
import unittest
from collections import UserDict

#
# Import local modules
#
from . import cache
from .test_server import HTTPTestBase, TestHTTPServer

#
# Module
#


class CacheTestBase(HTTPTestBase):

    CACHE_TTL = 30

    def setUp(self):
        HTTPTestBase.setUp(self)

        self.storage = self.getStorage()
        self.cache = cache.Cache(self.storage,
                                 timeToLiveSeconds=self.CACHE_TTL,
                                 userAgent='feedcache.test',
                                 )
        return

    def getStorage(self):
        "Return a cache storage for the test."
        return {}


class CacheTest(CacheTestBase):

    CACHE_TTL = 30

    def getServer(self):
        "These tests do not want to use the ETag or If-Modified-Since headers"
        return TestHTTPServer(applyModifiedHeaders=False)

    def testRetrieveNotInCache(self):
        # Retrieve data not already in the cache.
        feed_data = self.cache.fetch(self.TEST_URL)
        self.failUnless(feed_data)
        self.failUnlessEqual(feed_data.feed.title, 'CacheTest test data')
        return

    def testRetrieveIsInCache(self):
        # Retrieve data which is alread in the cache,
        # and verify that the second copy is identitical
        # to the first.

        # First fetch
        feed_data = self.cache.fetch(self.TEST_URL)

        # Second fetch
        feed_data2 = self.cache.fetch(self.TEST_URL)

        # Since it is the in-memory storage, we should have the
        # exact same object.
        self.failUnless(feed_data is feed_data2)
        return

    def testExpireDataInCache(self):
        # Retrieve data which is in the cache but which
        # has expired and verify that the second copy
        # is different from the first.

        # First fetch
        feed_data = self.cache.fetch(self.TEST_URL)

        # Change the timeout and sleep to move the clock
        self.cache.time_to_live = 0
        time.sleep(1)

        # Second fetch
        feed_data2 = self.cache.fetch(self.TEST_URL)

        # Since we reparsed, the cache response should be different.
        self.failIf(feed_data is feed_data2)
        return

    def testForceUpdate(self):
        # Force cache to retrieve data which is alread in the cache,
        # and verify that the new data is different.

        # Pre-populate the storage with bad data
        self.cache.storage[self.TEST_URL] = (time.time() + 100, self.id())

        # Fetch the data
        feed_data = self.cache.fetch(self.TEST_URL, force_update=True)

        self.failIfEqual(feed_data, self.id())
        return

    def testOfflineMode(self):
        # Retrieve data which is alread in the cache,
        # whether it is expired or not.

        # Pre-populate the storage with data
        self.cache.storage[self.TEST_URL] = (0, self.id())

        # Fetch it
        feed_data = self.cache.fetch(self.TEST_URL, offline=True)

        self.failUnlessEqual(feed_data, self.id())
        return

    def testUnicodeURL(self):
        # Pass in a URL which is unicode

        url = self.TEST_URL
        feed_data = self.cache.fetch(url)

        storage = self.cache.storage
        key = self.TEST_URL

        # Verify that the storage has a key
        self.failUnless(key in storage)

        # Now pull the data from the storage directly
        storage_timeout, storage_data = self.cache.storage.get(key)
        self.failUnlessEqual(feed_data, storage_data)
        return


class SingleWriteMemoryStorage(UserDict):
    """Cache storage which only allows the cache value
    for a URL to be updated one time.
    """

    def __setitem__(self, url, data):
        if url in self.keys():
            modified, existing = self[url]
            # Allow the modified time to change,
            # but not the feed content.
            if data[1] != existing:
                raise AssertionError('Trying to update cache for %s to %s' \
                                         % (url, data))
        UserDict.__setitem__(self, url, data)
        return


class CacheConditionalGETTest(CacheTestBase):

    CACHE_TTL = 0

    def getStorage(self):
        return SingleWriteMemoryStorage()

    def testFetchOnceForEtag(self):
        # Fetch data which has a valid ETag value, and verify
        # that while we hit the server twice the response
        # codes cause us to use the same data.

        # First fetch populates the cache
        response1 = self.cache.fetch(self.TEST_URL)
        self.failUnlessEqual(response1.feed.title, 'CacheTest test data')

        # Remove the modified setting from the cache so we know
        # the next time we check the etag will be used
        # to check for updates.  Since we are using an in-memory
        # cache, modifying response1 updates the cache storage
        # directly.
        response1['modified'] = None

        # This should result in a 304 status, and no data from
        # the server.  That means the cache won't try to
        # update the storage, so our SingleWriteMemoryStorage
        # should not raise and we should have the same
        # response object.
        response2 = self.cache.fetch(self.TEST_URL)
        self.failUnless(response1 is response2)

        # Should have hit the server twice
        self.failUnlessEqual(self.server.getNumRequests(), 2)
        return

    def testFetchOnceForModifiedTime(self):
        # Fetch data which has a valid Last-Modified value, and verify
        # that while we hit the server twice the response
        # codes cause us to use the same data.

        # First fetch populates the cache
        response1 = self.cache.fetch(self.TEST_URL)
        self.failUnlessEqual(response1.feed.title, 'CacheTest test data')

        # Remove the etag setting from the cache so we know
        # the next time we check the modified time will be used
        # to check for updates.  Since we are using an in-memory
        # cache, modifying response1 updates the cache storage
        # directly.
        response1['etag'] = None

        # This should result in a 304 status, and no data from
        # the server.  That means the cache won't try to
        # update the storage, so our SingleWriteMemoryStorage
        # should not raise and we should have the same
        # response object.
        response2 = self.cache.fetch(self.TEST_URL)
        self.failUnless(response1 is response2)

        # Should have hit the server twice
        self.failUnlessEqual(self.server.getNumRequests(), 2)
        return


class CacheRedirectHandlingTest(CacheTestBase):

    def _test(self, response):
        # Set up the server to redirect requests,
        # then verify that the cache is not updated
        # for the original or new URL and that the
        # redirect status is fed back to us with
        # the fetched data.

        self.server.setResponse(response, '/redirected')

        response1 = self.cache.fetch(self.TEST_URL)

        # The response should include the status code we set
        self.failUnlessEqual(response1.get('status'), response)

        # The response should include the new URL, too
        self.failUnlessEqual(response1.href, self.TEST_URL + 'redirected')

        # The response should not have been cached under either URL
        self.failIf(self.TEST_URL in self.storage)
        self.failIf(self.TEST_URL + 'redirected' in self.storage)
        return

    def test301(self):
        self._test(301)

    def test302(self):
        self._test(302)

    def test303(self):
        self._test(303)

    def test307(self):
        self._test(307)


class CachePurgeTest(CacheTestBase):

    def testPurgeAll(self):
        # Remove everything from the cache

        self.cache.fetch(self.TEST_URL)
        self.failUnless(self.storage.keys(),
                        'Have no data in the cache storage')

        self.cache.purge(None)

        self.failIf(self.storage.keys(),
                    'Still have data in the cache storage')
        return

    def testPurgeByAge(self):
        # Remove old content from the cache

        self.cache.fetch(self.TEST_URL)
        self.failUnless(self.storage.keys(),
                        'have no data in the cache storage')

        time.sleep(1)

        remains = (time.time(), copy.deepcopy(self.storage[self.TEST_URL][1]))
        self.storage['http://this.should.remain/'] = remains

        self.cache.purge(1)

        self.failUnlessEqual(list(self.storage), ['http://this.should.remain/'])
        return


if __name__ == '__main__':
    unittest.main()


feedcache/test_server.py
#!/usr/bin/env python
#
# Copyright 2007 Doug Hellmann.
#
#
#                         All Rights Reserved
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby
# granted, provided that the above copyright notice appear in all
# copies and that both that copyright notice and this permission
# notice appear in supporting documentation, and that the name of Doug
# Hellmann not be used in advertising or publicity pertaining to
# distribution of the software without specific, written prior
# permission.
#
# DOUG HELLMANN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
# NO EVENT SHALL DOUG HELLMANN BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
# OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""Simple HTTP server for testing the feed cache.

"""

__module_id__ = "$Id$"

#
# Import system modules
#
# import BaseHTTPServer
import logging

# import md5
import threading
import time
import unittest
import urllib.request

import hashlib
from http import server as BaseHTTPServer

#
# Import local modules
#


#
# Module
#
logger = logging.getLogger("feedcache.test_server")


def make_etag(data):
    """Given a string containing data to be returned to the client,
    compute an ETag value for the data.
    """
    _md5 = hashlib.md5()
    _md5.update(bytes(data, "utf-8"))
    return _md5.hexdigest()


class TestHTTPHandler(BaseHTTPServer.BaseHTTPRequestHandler):
    "HTTP request handler which serves the same feed data every time."

    FEED_DATA = """<?xml version="1.0" encoding="utf-8"?>

<feed xmlns="http://www.w3.org/2005/Atom" xml:lang="en-us">
  <title>CacheTest test data</title>
  <link href="http://localhost/feedcache/" rel="alternate"></link>
  <link href="http://localhost/feedcache/atom/" rel="self"></link>
  <id>http://localhost/feedcache/</id>
  <updated>2006-10-14T11:00:36Z</updated>
  <entry>
    <title>single test entry</title>
    <link href="http://www.example.com/" rel="alternate"></link>
    <updated>2006-10-14T11:00:36Z</updated>
    <author>
      <name>author goes here</name>
      <email>authoremail@example.com</email>
    </author>
    <id>http://www.example.com/</id>
    <summary type="html">description goes here</summary>
    <link length="100" href="http://www.example.com/enclosure" type="text/html" rel="enclosure">
    </link>
  </entry>
</feed>"""

    # The data does not change, so save the ETag and modified times
    # as class attributes.
    ETAG = make_etag(FEED_DATA)
    # Calculated using email.utils.formatdate(usegmt=True)
    MODIFIED_TIME = "Sun, 08 Apr 2012 20:16:48 GMT"

    def do_GET(self):
        "Handle GET requests."
        logger.debug("GET %s", self.path)

        if self.path == "/shutdown":
            # Shortcut to handle stopping the server
            logger.debug("Stopping server")
            self.send_response(200)
            self.end_headers()
            self.server.stop()

        else:
            # Record the request for tests that count them
            self.server.requests.append(self.path)
            # Process the request
            logger.debug("pre-defined response code: %d", self.server.response)
            handler_method_name = "do_GET_%d" % self.server.response
            handler_method = getattr(self, handler_method_name)
            handler_method()
        return

    def do_GET_3xx(self):
        "Handle redirects"
        if self.path.endswith("/redirected"):
            logger.debug("already redirected")
            # We have already redirected, so return the data.
            return self.do_GET_200()
        new_path = self.server.new_path
        logger.debug("redirecting to %s", new_path)
        self.send_response(self.server.response)
        self.send_header("Location", new_path)
        self.end_headers()
        return

    do_GET_301 = do_GET_3xx
    do_GET_302 = do_GET_3xx
    do_GET_303 = do_GET_3xx
    do_GET_307 = do_GET_3xx

    def do_GET_200(self):
        logger.debug("Etag: %s" % self.ETAG)
        logger.debug("Last-Modified: %s" % self.MODIFIED_TIME)

        incoming_etag = self.headers.get("If-None-Match", None)
        logger.debug('Incoming ETag: "%s"' % incoming_etag)

        incoming_modified = self.headers.get("If-Modified-Since", None)
        logger.debug("Incoming If-Modified-Since: %s" % incoming_modified)

        send_data = True

        # Does the client have the same version of the data we have?
        if self.server.apply_modified_headers:
            if incoming_etag == self.ETAG:
                logger.debug("Response 304, etag")
                self.send_response(304)
                self.end_headers()
                send_data = False

            elif incoming_modified == self.MODIFIED_TIME:
                logger.debug("Response 304, modified time")
                self.send_response(304)
                self.end_headers()
                send_data = False

        # Now optionally send the data, if the client needs it
        if send_data:
            logger.debug("Response 200")
            self.send_response(200)

            self.send_header("Content-Type", "application/atom+xml")

            logger.debug("Outgoing Etag: %s" % self.ETAG)
            self.send_header("ETag", self.ETAG)

            logger.debug("Outgoing modified time: %s" % self.MODIFIED_TIME)
            self.send_header("Last-Modified", self.MODIFIED_TIME)

            self.end_headers()

            logger.debug("Sending data")
            self.wfile.write(self.FEED_DATA.encode("utf-8"))
        return


class TestHTTPServer(BaseHTTPServer.HTTPServer):
    """HTTP Server which counts the number of requests made
    and can stop based on client instructions.
    """

    def __init__(self, applyModifiedHeaders=True, handler=TestHTTPHandler):
        self.apply_modified_headers = applyModifiedHeaders
        self.keep_serving = True
        self.requests = []
        self.setResponse(200)
        BaseHTTPServer.HTTPServer.__init__(self, ("", 9999), handler)
        return

    def setResponse(self, newResponse, newPath=None):
        """Sets the response code to use for future requests, and a new
        path to be used as a redirect target, if necessary.
        """
        self.response = newResponse
        self.new_path = newPath
        return

    def getNumRequests(self):
        "Return the number of requests which have been made on the server."
        return len(self.requests)

    def stop(self):
        "Stop serving requests, after the next request."
        self.keep_serving = False
        return

    def serve_forever(self):
        "Main loop for server"
        while self.keep_serving:
            self.handle_request()
        logger.debug("exiting")
        return


class HTTPTestBase(unittest.TestCase):
    "Base class for tests that use a TestHTTPServer"

    TEST_URL = "http://127.0.0.1:9999/"

    CACHE_TTL = 0

    def setUp(self):
        self.server = self.getServer()
        self.server_thread = threading.Thread(target=self.server.serve_forever)
        # set daemon flag so the tests don't hang if cleanup fails
        self.server_thread.setDaemon(True)
        self.server_thread.start()
        return

    def getServer(self):
        """Return a web server for the test."""
        s = TestHTTPServer()
        s.setResponse(200)
        return s

    def tearDown(self):
        # Stop the server thread
        urllib.request.urlretrieve(self.TEST_URL + "shutdown")
        time.sleep(1)
        self.server.server_close()
        self.server_thread.join()
        return


feedcache/test_shovefilesystem.py
#!/usr/bin/env python
#
# Copyright 2007 Doug Hellmann.
#
#
#                         All Rights Reserved
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby
# granted, provided that the above copyright notice appear in all
# copies and that both that copyright notice and this permission
# notice appear in supporting documentation, and that the name of Doug
# Hellmann not be used in advertising or publicity pertaining to
# distribution of the software without specific, written prior
# permission.
#
# DOUG HELLMANN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
# NO EVENT SHALL DOUG HELLMANN BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
# OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""Tests with shove filesystem storage.

"""

__module_id__ = "$Id$"

#
# Import system modules
#
import os
import shove
import tempfile
import unittest

#
# Import local modules
#
from .cache import Cache
from .test_server import HTTPTestBase

#
# Module
#

class CacheShoveTest(HTTPTestBase):

    def setUp(self):
        HTTPTestBase.setUp(self)
        self.shove_dirname = tempfile.mkdtemp('shove')
        return

    def tearDown(self):
        try:
            os.system('rm -rf %s' % self.storage_dirname)
        except AttributeError:
            pass
        HTTPTestBase.tearDown(self)
        return

    def test(self):
        # First fetch the data through the cache
        storage = shove.Shove('file://' + self.shove_dirname)
        try:
            fc = Cache(storage)
            parsed_data = fc.fetch(self.TEST_URL)
            self.failUnlessEqual(parsed_data.feed.title, 'CacheTest test data')
        finally:
            storage.close()

        # Now retrieve the same data directly from the shelf
        storage = shove.Shove('file://' + self.shove_dirname)
        try:
            modified, shelved_data = storage[self.TEST_URL]
        finally:
            storage.close()

        # The data should be the same
        self.failUnlessEqual(parsed_data, shelved_data)
        return


if __name__ == '__main__':
    unittest.main()


feedcache/example.py
#!/usr/bin/env python
#
# Copyright 2007 Doug Hellmann.
#
#
#                         All Rights Reserved
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby
# granted, provided that the above copyright notice appear in all
# copies and that both that copyright notice and this permission
# notice appear in supporting documentation, and that the name of Doug
# Hellmann not be used in advertising or publicity pertaining to
# distribution of the software without specific, written prior
# permission.
#
# DOUG HELLMANN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
# NO EVENT SHALL DOUG HELLMANN BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
# OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""Example use of feedcache.Cache.

"""

__module_id__ = "$Id$"

#
# Import system modules
#
import sys
import shelve

#
# Import local modules
#
import cache

#
# Module
#


def main(urls=[]):
    print("Saving feed data to ./.feedcache")
    storage = shelve.open(".feedcache")
    try:
        fc = cache.Cache(storage)
        for url in urls:
            parsed_data = fc.fetch(url)
            print(parsed_data.feed.title)
            for entry in parsed_data.entries:
                print("\t", entry.title)
    finally:
        storage.close()
    return


if __name__ == "__main__":
    main(sys.argv[1:])


feedcache/cache.py
#!/usr/bin/env python
#
# Copyright 2007 Doug Hellmann.
#
#
#                         All Rights Reserved
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby
# granted, provided that the above copyright notice appear in all
# copies and that both that copyright notice and this permission
# notice appear in supporting documentation, and that the name of Doug
# Hellmann not be used in advertising or publicity pertaining to
# distribution of the software without specific, written prior
# permission.
#
# DOUG HELLMANN DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN
# NO EVENT SHALL DOUG HELLMANN BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
# OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
# CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#

"""

"""

__module_id__ = "$Id$"

#
# Import system modules
#
import feedparser

import logging
import time

#
# Import local modules
#


#
# Module
#

logger = logging.getLogger("feedcache.cache")


class Cache:
    """A class to wrap Mark Pilgrim's Universal Feed Parser module
    (http://www.feedparser.org) so that parameters can be used to
    cache the feed results locally instead of fetching the feed every
    time it is requested. Uses both etag and modified times for
    caching.
    """

    def __init__(self, storage, timeToLiveSeconds=300, userAgent="feedcache"):
        """
        Arguments:

          storage -- Backing store for the cache.  It should follow
          the dictionary API, with URLs used as keys.  It should
          persist data.

          timeToLiveSeconds=300 -- The length of time content should
          live in the cache before an update is attempted.

          userAgent='feedcache' -- User agent string to be used when
          fetching feed contents.

        """
        self.storage = storage
        self.time_to_live = timeToLiveSeconds
        self.user_agent = userAgent
        return

    def purge(self, olderThanSeconds):
        """Remove cached data from the storage if the data is older than the
        date given.  If olderThanSeconds is None, the entire cache is purged.
        """
        if olderThanSeconds is None:
            logger.debug("purging the entire cache")
            for key in list(self.storage):
                del self.storage[key]
        else:
            now = time.time()
            # Iterate over the keys and load each item one at a time
            # to avoid having the entire cache loaded into memory
            # at one time.
            for url in list(self.storage):
                (cached_time, cached_data) = self.storage[url]
                age = now - cached_time
                if age >= olderThanSeconds:
                    logger.debug("removing %s with age %d", url, age)
                    del self.storage[url]
        return

    def fetch(self, url, force_update=False, offline=False):
        """Return the feed at url.

        url - The URL of the feed.

        force_update=False - When True, update the cache whether the
                                           current contents have
                                           exceeded their time-to-live
                                           or not.

        offline=False - When True, only return data from the local
                                 cache and never access the remote
                                 URL.

        If there is data for that feed in the cache already, check
        the expiration date before accessing the server.  If the
        cached data has not expired, return it without accessing the
        server.

        In cases where the server is accessed, check for updates
        before deciding what to return.  If the server reports a
        status of 304, the previously cached content is returned.

        The cache is only updated if the server returns a status of
        200, to avoid holding redirected data in the cache.
        """
        logger.debug('url="%s"' % url)

        # Convert the URL to a value we can use
        # as a key for the storage backend.
        key = url
        modified = None
        etag = None
        now = time.time()

        cached_time, cached_content = self.storage.get(key, (None, None))

        # Offline mode support (no networked requests)
        # so return whatever we found in the storage.
        # If there is nothing in the storage, we'll be returning None.
        if offline:
            logger.debug("offline mode")
            return cached_content

        # Does the storage contain a version of the data
        # which is older than the time-to-live?
        logger.debug("cache modified time: %s" % str(cached_time))
        if cached_time is not None and not force_update:
            if self.time_to_live:
                age = now - cached_time
                if age <= self.time_to_live:
                    logger.debug("cache contents still valid")
                    return cached_content
                else:
                    logger.debug("cache contents older than TTL")
            else:
                logger.debug("no TTL value")

            # The cache is out of date, but we have
            # something.  Try to use the etag and modified_time
            # values from the cached content.
            etag = cached_content.get("etag")
            modified = cached_content.get("modified")
            logger.debug("cached etag=%s" % etag)
            logger.debug("cached modified=%s" % str(modified))
        else:
            logger.debug("nothing in the cache, or forcing update")

        # We know we need to fetch, so go ahead and do it.
        logger.debug("fetching...")
        parsed_result = feedparser.parse(
            url,
            agent=self.user_agent,
            modified=modified,
            etag=etag,
        )

        status = parsed_result.get("status", None)
        logger.debug("HTTP status=%s" % status)
        if status == 304:
            # No new data, based on the etag or modified values.
            # We need to update the modified time in the
            # storage, though, so we know that what we have
            # stored is up to date.
            self.storage[key] = (now, cached_content)

            # Return the data from the cache, since
            # the parsed data will be empty.
            parsed_result = cached_content
        elif status == 200:
            # There is new content, so store it unless there was an error.
            error = parsed_result.get("bozo_exception")
            if not error:
                logger.debug("Updating stored data for %s" % url)
                self.storage[key] = (now, parsed_result)
            else:
                logger.warning("Not storing data with exception: %s", error)
        else:
            logger.warning("Not updating cache with HTTP status %s", status)

        return parsed_result
